### PR TITLE
Issue #91 - Ability to specify canaryService Service object to reach only canary pods

### LINKED ISSUE #91: Ability to specify canaryService Service object to reach only canary pods
Similar to blue-green's previewService, for users of the canary strategy, users would like to reach the pods of the canary through a service object, without hitting any of the old/stable pods. This proposal is to allow specifying a canary service like so:

```yaml
spec:
  strategy:
    canary:
      canaryService: guestbook-canary
```
If the user specifies a `canaryService` in the strategy, the controller will modify the specified service to add the `rollouts-pod-template-hash` as a label selector. 

For this to work, we will need to inject the `rollouts-pod-template-hash` label to the replicasets created by the canary strategy. This is something we already do for the blue-green strategy, but will need to do for canary to accomplish this feature.

### PR DESCRIPTION
**Pull request by alexmt**
Closes #91

### CODE REVIEW

File: controller/service.go [hunk 1]
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
...
+				c.persistRolloutStatus(r, newStatus, &r.Spec.Paused)
+			}
+			return err
+		}
+
+		if svc.Spec.Selector == nil {
+			svc.Spec.Selector = make(map[string]string)
+		}
+
+		hash := controller.ComputeHash(&r.Spec.Template, r.Status.CollisionCount)
```
Comment (dthomson25) [inline controller/service.go]: Instead of calculating the hash from the template, we should use calculate the hash from the newRS podSpec to protect against the k8s dependency change (#88).

File: controller/service.go [hunk 2]
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
...
+		if svc.Spec.Selector[v1alpha1.DefaultRolloutUniqueLabelKey] != hash {
+			patch := fmt.Sprintf(switchSelectorPatch, v1alpha1.DefaultRolloutUniqueLabelKey, hash)
```
Comment (dthomson25) [inline controller/service.go]: Can you use the c.switchServiceSelector method to patch the service?

File: controller/canary_test.go [hunk 3]
```diff
@@ -896,3 +898,52 @@ func TestCanaryRolloutStatusHPAStatusFields(t *testing.T) {
 	patch := f.getPatchedRollout(index)
 	assert.Equal(t, calculatePatch(r2, expectedPatchWithSub), patch)
 }
+
+func TestCanaryRolloutWithCanaryService(t *testing.T) {
+	f := newFixture(t)
+
+	canarySvc := newService("canary", 80, make(map[string]string))
```
Comment (dthomson25) [inline controller/canary_test.go]: What is there a reason that you instantiate the selector to nothing? If we don't, that will allow us to also test the case where the selector is nil.

File: controller/service.go [hunk 4]
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
...
+	if r.Spec.Strategy.CanaryStrategy != nil && r.Spec.Strategy.CanaryStrategy.CanaryService != "" {
+		svc, err := c.servicesLister.Services(r.Namespace).Get(r.Spec.Strategy.CanaryStrategy.CanaryService)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				msg := fmt.Sprintf(conditions.ServiceNotFoundMessage, r.Spec.Strategy.CanaryStrategy.CanaryService)
+				c.recorder.Event(r, corev1.EventTypeWarning, conditions.ServiceNotFoundReason, msg)
+				newStatus := r.Status.DeepCopy()
+				cond := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionFalse, conditions.ServiceNotFoundReason, msg)
+				conditions.SetRolloutCondition(newStatus, *cond)
+				c.persistRolloutStatus(r, newStatus, &r.Spec.Paused)
```
Comment (dthomson25) [inline controller/service.go]: I'm wondering if we should abstract this logic into a separate method because we have this code repeated 3 times. Thoughts?

File: controller/service.go [hunk 5]
```diff
@@ -234,6 +264,10 @@ func getRolloutServiceKeys(rollout *v1alpha1.Rollout) []string {
 		if rollout.Spec.Strategy.BlueGreenStrategy.PreviewService != "" {
 			servicesSet[fmt.Sprintf("%s/%s", rollout.Namespace, rollout.Spec.Strategy.BlueGreenStrategy.PreviewService)] = true
 		}
+	} else if rollout.Spec.Strategy.CanaryStrategy != nil {
```
Comment (dthomson25) [inline controller/service.go]: Can we add another test to TestGetRolloutServiceKeysForCanary to cover when a canaryService specified?

File: controller/canary_test.go (hunk 3 above)
Comment (alexmt) [reply controller/canary_test.go]: Good point. Changed `make(map[string]string)` to nil.

File: controller/service.go (hunk 5 above)
Comment (alexmt) [reply controller/service.go]: done

File: controller/service.go (hunk 1 above)
Comment (alexmt) [reply controller/service.go]: Done.

File: controller/service.go [hunk 6]
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
 	return previewSvc, activeSvc, nil
 }
 
+func (c *Controller) reconcileCanaryService(r *v1alpha1.Rollout, newRS *appsv1.ReplicaSet) error {
+	if r.Spec.Strategy.CanaryStrategy != nil && r.Spec.Strategy.CanaryStrategy.CanaryService != "" {
```
Comment (jessesuen) [inline controller/service.go]: Style nit, we would save a block of indentation with:
```golang
if r.Spec.Strategy.CanaryStrategy == nil || r.Spec.Strategy.CanaryStrategy.CanaryService == "" {
    return nil
}
```

File: controller/service.go (hunk 2 above)
Comment (alexmt) [reply controller/service.go]: done

File: controller/service.go (hunk 4 above)
Comment (alexmt) [reply controller/service.go]: good catch! done

File: controller/service.go (hunk 6 above)
Comment (alexmt) [reply controller/service.go]: done

**Comment by alexmt** [comment]
Addressed review comments. Also discovered that tests became flaky after merging https://github.com/argoproj/argo-rollouts/pull/95: additional service index requires run syncronize informers in test. The https://github.com/argoproj/argo-rollouts/pull/100/commits/9919549b147684e787414ddfb68c2e4f73f79941 fixes it

File: controller/service.go [hunk 7]
```diff
@@ -3,17 +3,16 @@ package controller
 import (
 	"fmt"
 
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/conditions"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
```
Comment (dthomson25) [inline controller/service.go]: Style nix: can we move this back below the external lib imports

Comment (alexmt) [reply controller/service.go]: done
